### PR TITLE
Added API to get player's proxy address

### PR DESCRIPTION
**Pull request by nostalfinals**
## Introduction

This patch added an API to get the player's proxy address.
It will be useful if the server uses multiple proxy instances that support HAProxy Protocol (frp, nginx, etc.) and wants to know which proxy instance the player uses.

## How?

I added a method, which is `Player#getHAProxyAddress`. This method returns an `InetSocketAddress`, if the server didn't enable Proxy Protocol, or the player didn't connect with an HAProxy instance, then it will return null.

## Example

```kotlin
@EventHandler
fun playerJoinEvent(event: PlayerJoinEvent) {
    val player = event.player
    val proxyAddr = player.haproxyAddress

    if (proxyAddr == null) {
        logger.info("Player ${player.name} didn't connect with a proxy!")
        return
    }

    logger.info("Player ${player.name} is using proxy: ${proxyAddr.hostString}:${proxyAddr.port}")
}
```

### CODE REVIEW

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+         this.receiving = side;
+diff --git a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
+index 586521a2cbb1d4dcfb912029f65e4363ec7674a7..44d26da371bd0b9ba970c24c85c62b5640120cd5 100644
+--- a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
++++ b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
+@@ -28,6 +28,7 @@ import io.netty.util.Timeout;
+ import io.netty.util.Timer;
+ import java.io.IOException;
+ import java.net.InetAddress;
++import java.net.InetSocketAddress;
```
Comment (powercasgamer) [inline patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]:
```suggestion

```

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+@@ -140,10 +141,17 @@ public class ServerConnectionListener {
+                                         String realaddress = message.sourceAddress();
+                                         int realport = message.sourcePort();
+ 
++                                        // Paper start - Add API to get player's proxy address
++                                        String proxyAddress = message.destinationAddress();
++                                        int proxyPort = message.destinationPort();
++
+                                         SocketAddress socketaddr = new java.net.InetSocketAddress(realaddress, realport);
++                                        SocketAddress proxyAddr = new InetSocketAddress(proxyAddress, proxyPort);
```
Comment (powercasgamer) [inline patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]:
```suggestion
+                                        SocketAddress proxyAddr = new java.net.InetSocketAddress(proxyAddress, proxyPort);
```

**Review by powercasgamer — CHANGES_REQUESTED** [review]
imports

File: patches/api/0469-Added-API-to-get-player-s-proxy-address.patch [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ 
++    // Paper start - Add API to get player's proxy address
++    /**
++    * Gets the socket address of this player's proxy
++    *
++    * @return the player's proxy address, null if the server didn't enable Proxy Protocol, or the player didn't connect with an HAProxy instance
++    */
++    @Nullable
++    public InetSocketAddress getProxyAddress();
++    // Paper end
```
Comment (notTamion) [inline patches/api/0469-Added-API-to-get-player-s-proxy-address.patch]: end comments also have identifiers

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+--- a/src/main/java/net/minecraft/network/Connection.java
++++ b/src/main/java/net/minecraft/network/Connection.java
+@@ -148,6 +148,10 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
+         this.stopReadingPackets = true;
+     }
+     // Paper end - packet limiter
++    // Paper start - Add API to get player's proxy address
++    @Nullable
++    public SocketAddress proxyAddress;
++    // Paper end
```
Comment (notTamion) [inline patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]: same here

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
++                                        String proxyAddress = message.destinationAddress();
++                                        int proxyPort = message.destinationPort();
++
+                                         SocketAddress socketaddr = new java.net.InetSocketAddress(realaddress, realport);
++                                        SocketAddress proxyAddr = new InetSocketAddress(proxyAddress, proxyPort);
+ 
+                                         Connection connection = (Connection) channel.pipeline().get("packet_handler");
+                                         connection.address = socketaddr;
++                                        connection.proxyAddress = proxyAddr;
++                                        // Paper end
```
Comment (notTamion) [inline patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]: same here + you can move your stuff below all of mojangs code so you don't also have all that vanilla code between your comments.

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
++
++        SocketAddress addr = this.getHandle().connection.connection.proxyAddress;
++
++        if (addr instanceof InetSocketAddress) {
++            return (InetSocketAddress) addr;
++        }
++
++        return null;
++    }
++    // Paper end
```
Comment (notTamion) [inline patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]: same here

**Review by notTamion — CHANGES_REQUESTED** [review]

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch (hunk 5 above)
Comment (notTamion) [reply patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]: i think i should also just add here that we probably don't necessarily need to create all those variables to match mojangs style here.

File: patches/api/0469-Added-API-to-get-player-s-proxy-address.patch [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -248,6 +248,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     @Nullable
+     public InetSocketAddress getAddress();
+ 
++    // Paper start - Add API to get player's proxy address
++    /**
++    * Gets the socket address of this player's proxy
++    *
++    * @return the player's proxy address, null if the server didn't enable Proxy Protocol, or the player didn't connect with an HAProxy instance
```
Comment (powercasgamer) [inline patches/api/0469-Added-API-to-get-player-s-proxy-address.patch]:
```suggestion
+    * @return the player's proxy address, null if the server doesn't have Proxy Protocol enabled, or the player didn't connect to an HAProxy instance
```
Something like this?

File: patches/server/1055-Added-API-to-get-player-s-proxy-address.patch [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+--- a/src/main/java/net/minecraft/network/Connection.java
++++ b/src/main/java/net/minecraft/network/Connection.java
+@@ -148,6 +148,10 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
+         this.stopReadingPackets = true;
+     }
+     // Paper end - packet limiter
++    // Paper start - Add API to get player's proxy address
++    @Nullable
++    public SocketAddress proxyAddress;
++    // Paper end - Add API to get player's proxy address
```
Comment (powercasgamer) [inline patches/server/1055-Added-API-to-get-player-s-proxy-address.patch]:
```suggestion
+    @Nullable
+    public SocketAddress proxyAddress; // Paper - Add API to get player's proxy address
```

**Review by powercasgamer — CHANGES_REQUESTED** [review]

File: patches/api/0469-Added-API-to-get-player-s-proxy-address.patch [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+     public InetSocketAddress getAddress();
+ 
++    // Paper start - Add API to get player's proxy address
++    /**
++    * Gets the socket address of this player's proxy
++    *
++    * @return the player's proxy address, null if the server doesn't have Proxy Protocol enabled, or the player didn't connect to an HAProxy instance
++    */
++    @Nullable
++    public InetSocketAddress getProxyAddress();
```
Comment (electronicboy) [inline patches/api/0469-Added-API-to-get-player-s-proxy-address.patch]: This should generally be more declarative to what this is actually exposing, as it generally feels detached from what it actually exposes

**Review by lynxplay — COMMENTED** [review]
The naming here feels a bit off.
I don#t have a better suggestion at hand, but given how rare the usage of HA proxies infront of servers are, I feel like the name `proxyAddress` does not properly separate that this method is exlusivly for HA proxies and not normal "proxies" as the minecraft community would understand (e.g. velo/waterfall/bungee).

File: patches/api/0469-Added-API-to-get-player-s-proxy-address.patch (hunk 9 above)
Comment (yannicklamprecht) [reply patches/api/0469-Added-API-to-get-player-s-proxy-address.patch]: getProxyJoinedWithAddress() or is that even worse?

Comment (nostalfinals) [reply patches/api/0469-Added-API-to-get-player-s-proxy-address.patch]: I'm considering if just name it as `getHAProxyAddress` directly will be better.

Comment (lynxplay) [reply patches/api/0469-Added-API-to-get-player-s-proxy-address.patch]: I like it too 👍

**Comment by nostalfinals** [comment]
I've already changed the method name and updated this to 1.20.5.

File: patches/server/1046-Added-API-to-get-player-s-proxy-address.patch [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
++    // Paper start - Add API to get player's proxy address
++    @Override
++    public @Nullable InetSocketAddress getHAProxyAddress() {
++        if (this.getHandle().connection == null) {
++            return null;
++        }
++
++        SocketAddress addr = this.getHandle().connection.connection.proxyAddress;
++
++        if (addr instanceof InetSocketAddress) {
```
Comment (kennytv) [inline patches/server/1046-Added-API-to-get-player-s-proxy-address.patch]: pattern matching

**Review by kennytv — APPROVED** [review]

File: patches/server/1046-Added-API-to-get-player-s-proxy-address.patch (hunk 10 above)
Comment (yannicklamprecht) [reply patches/server/1046-Added-API-to-get-player-s-proxy-address.patch]:
```java
return this.getHandle().connection.connection.proxyAddress instanceof InetSocketAddress inetSocketAddress ? inetSocketAddress : null;
```


or alternative keeping the GuardClause.

```java
    @Override
    public @Nullable InetSocketAddress getHAProxyAddress() {
        if (this.getHandle().connection == null || !(this.getHandle().connection.connection.proxyAddress instanceof InetSocketAddress inetSocketAddress)) {
            return null;
        }
        return inetSocketAddress;
    }

 ```

**Review by lynxplay — APPROVED** [review]

**Review by Machine-Maker — APPROVED** [review]

**Pull request merged**